scheds/rust/scx_mitosis/src/bpf/mitosis.bpf.c

```diff
@@ -0,0 +1,932 @@
+		}
+		bpf_cgroup_release(cgrp);
+	}
+
+	if (!(cell_cpumask = (struct cpumask *)tctx->cell_cpumask) ||
+	    !(idle_smtmask = scx_bpf_get_idle_smtmask()))
+		return prev_cpu;
+
+	bpf_cpumask_and((struct bpf_cpumask *)cell_cpumask, cell_cpumask,
+			p->cpus_ptr);
+	/* No overlap between cell cpus and task cpus, just send it to global */
+	if (bpf_cpumask_empty(cell_cpumask)) {
+		cstat_inc(CSTAT_AFFN_VIOL, tctx->cell, cctx);
+
+		if ((cpu = pick_idle_cpu_from(p->cpus_ptr, prev_cpu,
+					      idle_smtmask)) >= 0)
+			goto dispatch_local;
+
+		cstat_inc(CSTAT_GLOBAL, tctx->cell, cctx);
+		scx_bpf_dispatch(p, SCX_DSQ_GLOBAL, slice_ns, 0);
+		scx_bpf_put_idle_cpumask(idle_smtmask);
+		return prev_cpu;
+	}
+
+	if ((cpu = pick_idle_cpu_from(cell_cpumask, prev_cpu, idle_smtmask)) >=
+	    0)
+		goto dispatch_local;
+
+	if (bpf_cpumask_test_cpu(prev_cpu, cell_cpumask))
+		cpu = prev_cpu;
+	else
+		cpu = bpf_cpumask_any_distribute(cell_cpumask);
+	goto out_put_idle_smtmask;
+
+dispatch_local:
+	local = true;
+	cstat_inc(CSTAT_LOCAL, tctx->cell, cctx);
+	scx_bpf_dispatch(p, SCX_DSQ_LOCAL, slice_ns, 0);
+out_put_idle_smtmask:
+	scx_bpf_put_idle_cpumask(idle_smtmask);
+	if (debug && !draining && tctx->all_cpus_allowed &&
+	    (cctx = lookup_cpu_ctx(cpu)) && cctx->cell != tctx->cell)
+		scx_bpf_error(
+			"select_cpu returned cpu %d belonging to cell %d but task belongs to cell %d, local %d",
+			cpu, cctx->cell, tctx->cell, local);
+	return cpu;
+}
+
+static inline bool vtime_before(u64 a, u64 b)
+{
+	return (s64)(a - b) < 0;
+}
+
+void BPF_STRUCT_OPS(mitosis_enqueue, struct task_struct *p, u64 enq_flags)
+{
+	struct cpu_ctx *cctx;
+	struct task_ctx *tctx;
+	struct cell *cell;
+	u64 vtime = p->scx.dsq_vtime;
+
+	if (!(cctx = lookup_cpu_ctx(-1)) || !(tctx = lookup_task_ctx(p)) ||
+	    !(cell = lookup_cell(tctx->cell)))
+		return;
+
+	/*
+	 * Limit the amount of budget that an idling task can accumulate
+	 * to one slice.
+	 */
+	if (vtime_before(vtime, cell->vtime_now - slice_ns))
+		vtime = cell->vtime_now - slice_ns;
+
+	scx_bpf_dispatch_vtime(p, tctx->cell, slice_ns, vtime, enq_flags);
+}
+
+void BPF_STRUCT_OPS(mitosis_dispatch, s32 cpu, struct task_struct *prev)
+{
+	struct cpu_ctx *cctx;
+	u32 prev_cell, cell;
+
+	if (!(cctx = lookup_cpu_ctx(-1)))
+		return;
+
+	prev_cell = *(volatile u32 *)&cctx->prev_cell;
+	cell = *(volatile u32 *)&cctx->cell;
+
+	/*
+	 * cpu <=> cell assignment can change dynamically. In order to deal with
+	 * scheduling racing with assignment change, we schedule from the previous
+	 * cell first to make sure it drains.
+	 */
+	if (prev_cell != cell && scx_bpf_consume(prev_cell))
+		return;
+
+	scx_bpf_consume(cell);
+}
+
+static inline void runnable(struct task_struct *p, struct task_ctx *tctx,
+			    struct cgroup *cgrp)
+{
+	struct cgrp_ctx *cgc;
+
+	if (tctx->cell == -1) {
+		if (!(cgc = lookup_cgrp_ctx(cgrp)))
+			return;
+
+		tctx->cell = cgc->cell;
+	}
+
+	adj_load(p, tctx, cgrp, p->scx.weight, bpf_ktime_get_ns());
```

Comment:
just FYI, if you end up wanting to support infeasible weights in user space, you'll probably end up having to track duty cycle here instead of load, and do the scaling in user space when you do an lb pass